### PR TITLE
feat: add background options and borders to styling panel

### PR DESCRIPTION
**Pull request by Donya-qlik**
The background and border styling options have been added to the action-button chart styling panel, it includes the listed items.  

- Background color
- Background Image using MediaLibrary
- Border
- Reset the section

Feature-flag: SENSECLIENT_IM_1525_STYLINGPANEL_BUTTON
Note: To be able to test it locally, sense-client needs to be linked to analytic changes on my branch https://github.com/qlik-trial/analytics/pull/401 and for getting new translations to https://github.com/qlik-trial/sense-client/pull/23139

Unsolved Issues of the current PR:

- Color Picker preview boxes are not showing the default colors that should be fixed in the coming PRs.
- The position grid does not show focus on center at the beginning. It is because the old and new version of positionGrid component take different values for the default position. The current value is `centerCenter` whereas it is 'center-center' with the new component. 

https://user-images.githubusercontent.com/20701546/219906767-1bd6c05c-8334-4962-8fe2-adcfafdacf49.mp4

### CODE REVIEW

File: src/utils/style-utils.js [hunk 1]
```diff
@@ -48,3 +81,19 @@ export const fontFamilyOptions = FONT_FAMILIES.map((font) => ({
   value: font,
   label: getFirstFont(font),
 }));
+
+export const getColor = (
```
Comment (haxxmaxx) [inline src/utils/style-utils.js:85]: since there is a color-utils, doesn't this fit better there?

File: src/utils/style-formatter.js [hunk 2]
```diff
@@ -2,71 +2,39 @@ import colorUtils from './color-utils';
...
     const fontStyle = font.style || DEFAULTS.FONT_STYLE;
     fontStyle.bold && (styles += formatProperty('font-weight', 'bold'));
     fontStyle.italic && (styles += formatProperty('font-style', 'italic'));
     // background
+    const primaryColor = theme.getDataColorSpecials().primary;
+    const backgroundColor = getColor(background, primaryColor, theme);
     styles += formatProperty('background-color', backgroundColor);
-    if (background.useImage && background.url.qStaticContentUrl) {
+    // backgroundImage
+    const {useImage} = background;
```
Comment (haxxmaxx) [inline src/utils/style-formatter.js:26]: Are you using the linter? It should do this automatically
```suggestion
    const { useImage } = background;
```

File: src/object-properties.js [hunk 3]
```diff
@@ -159,12 +160,13 @@ const properties = {
  * @property {boolean=} [useColorExpression=false] - Set to true to use color expression
  * @property {PaletteColor=} color - Color defined by index or hex code, needed if useColorExpression is false
  * @property {ColorExpression=} colorExpression - Color defined by expression, needed if useColorExpression is true
+ * @property {string=} [mode = 'none'] - Set to media library to show background image
  * @property {boolean=} [useImage=false] - Set to true to show background image
```
Comment (haxxmaxx) [inline src/object-properties.js:164]: wait, do we need both here? looks like useImage and mode does the same thing

**Review by haxxmaxx — APPROVED** [review]

**Comment by haxxmaxx** [comment]
woops didn't mean to approve, seems like I can't revert it. Will approve again when comments are taken care of

File: src/utils/style-utils.js (hunk 1 above)
Comment (Donya-qlik) [reply src/utils/style-utils.js:85]: I found color-utils it more related to set css colors and color code stuff... but sure I can move it over there.

File: src/object-properties.js (hunk 3 above)
Comment (Donya-qlik) [reply src/object-properties.js:174]: Well...the old component of use-background-image in the main panel(not styling one) is a switch that here it is defined by a boolean useImage. The new one is a dropdown that is called mode, and has two options to show/hide the image, so should keep both.

File: src/utils/style-utils.js [hunk 4]
```diff
@@ -21,6 +24,36 @@ const FONT_FAMILIES = [
...
+export const backgroundPosition = {
+  'top-left': 'top left',
+  'center-left': 'center left',
+  'bottom-left': 'bottom left',
+  'top-center': 'top center',
+  'center-center': 'center center',
+  'bottom-center': 'bottom center',
+  'top-right': 'top right',
+  'center-right': 'center right',
+  'bottom-right': 'bottom right',
```
Comment (Donya-qlik) [inline src/utils/style-utils.js:42]: @niekvanstaveren I tried with replacing the these values with percentage ones (e.x: 50% 50%) but it didn't work, clicking around position component didn't change the image position so I kept it as it is.

File: src/styling-panel-definition.js [hunk 5]
```diff
@@ -52,14 +52,17 @@ export const getStyleEditorDefinition = () => ({
                   type: 'boolean',
                   translation: 'properties.fontColor',
                   component: 'dropdown',
+                  defaultValue: false,
```
Comment (niekvanstaveren) [inline src/styling-panel-definition.js:55]: I think we made a mistake here in the past. A dropdown and boolean should not be together

File: src/styling-panel-definition.js [hunk 6]
```diff
@@ -77,6 +80,191 @@ export const getStyleEditorDefinition = () => ({
...
+                  value: 'media',
+                  translation: 'MediaLibrary.Header',
+                },
+              ],
+              change(data) {
+                const bgImageComp = data.style.background;
+                bgImageComp
+                  ? (bgImageComp.url = { qStaticContentUrlDef: '' })
+                  : (data.style.background = { url: { qStaticContentUrlDef: '' } });
+              },
```
Comment (niekvanstaveren) [inline src/styling-panel-definition.js:139]: Why resetting the background image here?

File: src/styling-panel-definition.js [hunk 7]
```diff
@@ -77,6 +80,191 @@ export const getStyleEditorDefinition = () => ({
...
+                  translation: 'properties.backgroundImage.sizeStretch',
+                },
+                {
+                  value: 'alwaysFill',
+                  translation: 'properties.backgroundImage.sizeAlwaysFill',
+                },
+              ],
+              change: (data) => {
+                let currentPosition = propertyResolver.getValue(data, 'style.background.position');
+                if (currentPosition) currentPosition = styleDefaults.BGIMAGE_POSITION;
```
Comment (niekvanstaveren) [inline src/styling-panel-definition.js:179]: Do we need to reset the position every time we change the size?

File: src/styling-panel-definition.js [hunk 8]
```diff
@@ -77,6 +80,191 @@ export const getStyleEditorDefinition = () => ({
...
+      translation: 'properties.border',
+      items: {
+        backgroundBorderItem: {
+          items: {
+            useBorder: {
+              type: 'boolean',
+              component: 'switch',
+              ref: 'style.border.useBorder',
+              translation: 'properties.border.use',
+              defaultValue: toggleOptions[1].value,
```
Comment (niekvanstaveren) [inline src/styling-panel-definition.js:211]: Just `false` would be easier I think

**Review by niekvanstaveren — APPROVED** [review]
Some small comments. Please have a look at them

File: src/utils/__tests__/style-formatter.spec.js [hunk 9]
```diff
@@ -128,7 +192,7 @@ describe('style-formatter', () => {
           },
         },
       };
-      const formattedStyle = styleFormatter.getStyles({ style, disabled, theme, app });
+      let formattedStyle = styleFormatter.getStyles({ style, disabled, theme, app });
```
Comment (niekvanstaveren) [inline src/utils/__tests__/style-formatter.spec.js:195]: This should be a const (the lint is also complaining about it)

File: src/styling-panel-definition.js (hunk 5 above)
Comment (Donya-qlik) [reply src/styling-panel-definition.js:55]: Ah I see I forgot to remove it from here, I did in the later PRs. I'll fix it here.

File: src/styling-panel-definition.js (hunk 7 above)
Comment (Donya-qlik) [reply src/styling-panel-definition.js:179]: Changing happens when you select another size mode and each option enables/disables some positions spots. Thus, yes it puts it back to center and let you choose the new position.

File: src/styling-panel-definition.js (hunk 8 above)
Comment (Donya-qlik) [reply src/styling-panel-definition.js:211]: Oh weird, I had fixed some details before but they are still here... I'll fix it.

**Pull request merged**